src/treekem.cpp

```diff
@@ -869,6 +876,81 @@ TreeKEMPublicKey::parent_hashes(
   return ph;
 }
 
+const bytes&
+TreeKEMPublicKey::original_tree_hash(TreeHashCache& cache,
+                                     NodeIndex index,
+                                     std::vector<LeafIndex> parent_except) const
+{
+  // Scope the unmerged leaves list down to this subtree
+  auto except = std::vector<LeafIndex>{};
+  std::copy_if(parent_except.begin(),
+               parent_except.end(),
+               std::back_inserter(except),
+               [&](auto i) { return NodeIndex(i).is_below(index); });
+
+  auto have_local_changes = !except.empty();
+
+  // If there are no local changes, then we can use the cached tree hash
+  if (!have_local_changes) {
+    return hashes.at(index);
+  }
+
+  // If this method has been called before with the same number of excluded
+  // leaves (which implies the same set), then use the cached value.
+  auto cache_key = std::make_pair(index, except.size());
+  if (auto it = cache.find(cache_key); it != cache.end()) {
```

Comment:
just curious, is it possible that same sized excluded set can happen but with different operations 
on the tree between the calls here ?
I want to make sure the excluded set size check is a safe assumption

Reply:
Good question.  That came from some code @TWal added as we were prototyping this.  I believe the argument goes as follows:

* The possible `except` sets here are the intersection of the `unmerged_leaves` of ancestor nodes with the current subtree
* Because nodes always get reset along a direct path, if a leaf is unmerged at a node `P`, then it unmerged at all the non-blank nodes between the leaf and `P`.
* As a result, the `except` sets relative to ancestors are all subsets of one another.  If there are ancestors `P1`, ..., `Pk` from the current node to the root with corresponding `except` sets `E1, ..., Ek`, then `E1 >= E2 >= ... >= Ek`.
* That means that among the possible `exclude` sets, there's only one with a given size.

Reply:
Yes, this is the argument indeed!

However, when working with parent-hash security proofs, I realized there are two kind of invariants:
- the ones true for group creation (base case), and group modification (induction case)
- the ones true for group creation AND accepting a Welcome message (base case), and group modifications (induction case)

The correctness of this implementation relies on an invariant of the first kind, which is not checked when accepting a Welcome message.
This implementation won't accept trees that are rejected by the RFC, however it might reject trees that are accepted by the RFC (in this case it means that someone sent you a bogus tree, so things are weird anyway…)

Something we could do is to add to the Welcome check all the invariants of MLS, so that if the Welcome check succeeds, there exists a sequence of operations that starting from a group creation, that gives the tree you are processing. Hence, the first kind of invariants and the second kind of invariants would be the same.